core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2937,7 +2937,9 @@ private[spark] class DAGScheduler(
         } else {
           // This stage is only used by the job, so finish the stage if it is running.
           val stage = stageIdToStage(stageId)
-          if (runningStages.contains(stage)) {
+          val isRunningStage = runningStages.contains(stage) ||
+            (waitingStages.contains(stage) && taskScheduler.hasRunningTasks(stageId))
```

Comment:
what if we just kill all waiting stages? Does `taskScheduler.killAllTaskAttempts` handle it well?

Reply:
and can we add a special flag to indicate the waiting stages that are submitted due to retry?

Reply:
> what if we just kill all waiting stages? Does taskScheduler.killAllTaskAttempts handle it well?
I tested that: if the normally generated waiting stages call `killAllTaskAttempts`,  the stage status will be displayed as `FAILED`, which was `SKIPPED` before, killAllTaskAttempts  itself will not go wrong.
Always kill waiting stage seems to be a safer approach (tasks shouldn't be run anymore), but it may generate unnecessary `stageFailed` events compared with before.

> and can we add a special flag to indicate the waiting stages that are submitted due to retry?
Yes, we can add a flag , please see the update codes. 
Actually , there is a trade-off here to kill waiting stages
- always kill
- kill stages had failed ( `stage#failedAttemptIds` > 0)
- kill stages failed when fetch failed (`stage#resubmitInFetchFailed`)
- kill stages only have running tasks